File: Data Structures/java/stack/linked/LinkedStack.java
package stack.linked;

import stack.Stack;
import stack.StackEmptyException;

public class LinkedStack implements Stack {
    private Node top;
    private int count;

    public LinkedStack() {
        top = null;
        count = 0;
    }

    public LinkedStack(Object obj) {
        top = new Node(obj);
        count = 1;
    }

    public int size() {
        return count;
    }

    public boolean isFull() {
        return false;
    }

    public boolean isEmpty() {
        return count == 0;
    }

    public void push(Object obj) throws StackEmptyException {
        Node temp = new Node(obj);
        temp.next = top;
        top = temp;
        count++;
    }

    public Object pop() throws StackEmptyException {
        if (isEmpty()) {
            throw new StackEmptyException("Stack is empty");
        }
        Object obj = top.data;
        top = top.next;
        count--;
        return obj;
    }

    public Object peek() throws StackEmptyException {
        if (isEmpty()) {
            throw new StackEmptyException("Stack is empty");
        }
        return top.data;
    }
}

File: Data Structures/c/stack/LinkedStack.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node Node;

struct Node {
    int data;
    Node *next;
};

int isEmpty();
int isFull();
void peek();
void push(int);
void pop();
void print();

Node* top;
int count = 0;

int main() {
    push(1);
    push(2);
    push(3);
    push(4);
    peek();
    pop();
    pop();
    print();
    push(8);
    print();
    return 0;
}

int isEmpty() {
    return count == 0;
}

void peek() {
    if (isEmpty()) {
        printf("Stack kosong\n");
    } else {
        printf("Top stack adalah %d\n", top->data);
    }
}

void push (int data) {
    Node* temp = malloc(sizeof(Node));
    temp->data = data;
    temp->next = top;
    top = temp;
    count++;
}

void pop () {
    if (isEmpty()) {
        printf("Stack kosong\n");
    } else {
        printf("Data yang dihapus adalah %d\n", top->data);
        top = top->next;
        count--;
    }
}

void print() {
    if (isEmpty()) {
        printf("Stack kosong\n");
    } else {
        Node* temp = top;
        while(temp != NULL) {
            printf("%d ", temp->data);
            temp = temp->next;
        }
        printf("\n");
    }
}